Transfer-Function ITS: Graded Interventions with Saturation & Adstock Transforms

drbenvincent opened the pull request:
This PR introduces Transfer-Function Interrupted Time Series (TF-ITS), a powerful new experiment class that extends CausalPy's causal inference capabilities to handle graded (non-binary) interventions in single-market time series data.

One reason why it's exciting is that it starts to create a bridge between CausalPy and pymc-marketing.

## 🎯 What This Adds

Unlike traditional ITS methods that focus on binary on/off treatments, TF-ITS enables practitioners to:
- Model interventions with varying intensity (e.g., advertising spend, policy intensity, promotional campaigns)
- Account for diminishing returns through saturation transforms (Hill, logistic, Michaelis-Menten)
- Capture carryover effects using adstock transforms (geometric decay with configurable half-life)
- Estimate window-level causal lift by constructing counterfactuals with scaled or zeroed interventions

This makes TF-ITS particularly valuable for marketing mix modeling, policy evaluation, and any scenario where treatment intensity varies over time.

## 🔧 Implementation Details

- Estimation: OLS + Bayesian.
- Transforms: Leverages pymc-marketing transform library for saturation and adstock functions
- Counterfactuals: Flexible engine for computing treatment effects by scaling exposures in specified time windows
- Diagnostics: Residual ACF/PACF plots and Ljung-Box tests
- Visualization: Model fit plots and impulse response functions (IRF)

Architecture:

- Modular transform system with dataclasses (Saturation, Adstock, Lag, Treatment). Very open to discussing if this is the best way forward or not.
- Clean separation between transform specification and application logic

---

Dependencies: Adds pymc-marketing>=0.7.0
Breaking Changes: None
Future Work: Multiple intervention channels (this may be close to working, but just not worked through an example). Add another notebook directly focussing on a marketing-based case study.

<!-- readthedocs-preview causalpy start -->
----
📚 Documentation preview 📚: https://causalpy--548.org.readthedocs.build/en/548/

<!-- readthedocs-preview causalpy end -->

drbenvincent committed:
very initial MVP

drbenvincent committed:
Add notebook to the docs page so it will now render in the docs build

drbenvincent committed:
change example + add cell tags

drbenvincent committed:
add properly formatted reference section

drbenvincent committed:
Refactor transform strategy and add parameter estimation

Refactored transform classes to use a strategy pattern with explicit Adstock, Saturation, and Lag implementations. Added transform_optimization.py for grid search and optimization of transform parameters. Updated TransferFunctionITS to support transform parameter estimation and metadata. Revised tests to use new transform classes and parameter estimation workflows.

drbenvincent commented on causalpy/transforms.py:
```diff
@@ -0,0 +1,416 @@
+
+
+@dataclass
+class Saturation:
```
Hopefully resolved in https://github.com/pymc-labs/CausalPy/pull/548/commits/659b5020205d43a0e978f3406b795b234d85043e

drbenvincent committed:
Improve TF-ITS docs and parameter estimation details

Expanded and clarified docstrings in transfer_function_its.py to document the nested parameter estimation approach for saturation and adstock transforms. Updated the example and usage instructions to reflect the new estimation workflow. Revised the notebook to demonstrate transform parameter estimation via grid search, show parameter recovery, and clarify the distinction between grid search and continuous optimization. Removed the outdated and redundant test class for TransferFunctionITS in test_transfer_function_its.py.

drbenvincent committed:
get tests passing

drbenvincent committed:
fix doctest errors

drbenvincent committed:
Clarify HAC standard errors in docs and code

Expanded documentation and code comments to better explain HAC (Newey-West) standard errors, their purpose, and the hac_maxlags parameter. Added a detailed explanation and citation in the notebook, and improved docstrings and print output in transfer_function_its.py. Added the Newey-West reference to references.bib.

drbenvincent committed:
Add section on autocorrelation and HAC standard errors

Expanded the TF-ITS notebook with a detailed explanation of autocorrelation in time series, its impact on causal inference, and the motivation for using HAC (Newey-West) standard errors. Updated the simulation to generate autocorrelated errors using an AR(1) process, and clarified the importance of robust inference in the context of time series interventions.

drbenvincent committed:
move reference section to the bottom where it should be

drbenvincent committed:
Add ARIMAX error model support to TransferFunctionITS

Extended TransferFunctionITS and transform optimization to support ARIMAX (ARIMA with exogenous variables) error models in addition to HAC standard errors. Updated model fitting, parameter estimation, and documentation to allow users to specify error_model ('hac' or 'arimax') and ARIMA order. Added comprehensive tests for ARIMAX functionality and updated the notebook to demonstrate ARIMAX usage and comparison with HAC.

drbenvincent committed:
refactor into causalpy class structure (not updated notebook yet)

drbenvincent committed:
Refactor graded intervention ITS to use unfitted model pattern

Refactors GradedInterventionTimeSeries and TransferFunctionOLS to follow the standard CausalPy pattern: the experiment class now takes an unfitted model and handles transform parameter estimation, fitting, and result extraction. Removes the with_estimated_transforms factory method, updates all docstrings, and adapts tests and documentation to the new workflow. This enables more flexible and consistent usage for multi-treatment and advanced modeling scenarios.

drbenvincent committed:
Add plotting utilities and improve docs for transfer function ITS

Introduces new plotting methods to GradedInterventionTimeSeries, including plot_effect and plot_transforms, and renames diagnostics() to plot_diagnostics(). Updates tests to cover new plotting features. Enhances documentation and notebook explanations for model fitting and parameter estimation, and updates the interrogate badge.

drbenvincent committed:
fix error that was making doctests fail

drbenvincent committed:
re-brand to graded intervention time series

Renamed 'tfits_single_channel.ipynb' to 'graded_intervention_time_series_single_channel_ols.ipynb' and updated the notebook title and references in both the notebook and the index.md file to reflect the new name and description.

drbenvincent committed:
remove implementation summary

drbenvincent committed:
minor docs updates

drbenvincent committed:
minor restructure of notebook for clarity, readability, and flow

drbenvincent committed:
more tests to address failing code coverage check

drbenvincent committed:
add to complexity/realism of simulated data

drbenvincent committed:
add mention of multiple treatments + citation

drbenvincent committed:
add reference + citation to helfenstein1991use

drbenvincent committed:
change notebook title

drbenvincent committed:
update wording around transfer functions

drbenvincent committed:
update wording around error model

drbenvincent committed:
minot formatting updates

drbenvincent committed:
same basic structure for each of the modelling sections

drbenvincent committed:
rerun notebook

drbenvincent committed:
Support adstock-only and saturation-only transfer functions

Refactored transfer function parameter estimation and plotting to allow models with only adstock or only saturation transforms (or both). Updated validation logic, grid/optimize routines, and plotting to handle optional transforms. Added comprehensive tests for all transform configurations and clarified documentation and notebook to demonstrate adstock-only use case.

drbenvincent committed:
first stab at demo of optimization

drbenvincent committed:
improve plotting of adstock functions

drbenvincent committed:
reduce the causal impact for more realism

drbenvincent committed:
First stab at adding Bayesian transfer function model for graded ITS

drbenvincent committed:
progressing the Bayesian implementation

drbenvincent committed:
initial attempt at Bayesian AR(1) errors model

drbenvincent committed:
better priors scaled for the problem

drbenvincent committed:
update to use `Prior` class from pymc-extras + fix failing doctests

drbenvincent committed:
Expand notes on AR error modeling in TransferFunctionLinearRegression

Added detailed documentation explaining challenges of AR error modeling in PyMC, why standard approaches fail, and the rationale for using quasi-differencing in TransferFunctionARRegression. Also clarified alternative latent AR component modeling and why it is not used, providing guidance on when to use each model.

drbenvincent committed:
fix doctests

drbenvincent committed:
Use dedicated sampling parameters for AR model test

Introduces a separate 'ar_sample_kwargs' dictionary for Bayesian AR(1) model sampling in the transfer function test. Updates assertions to reference the new parameters, clarifying the need for increased sampling due to model complexity.

drbenvincent committed:
sort test

drbenvincent committed:
increase test coverage

drbenvincent committed:
increase code coverage

drbenvincent committed:
more test coverage

drbenvincent committed:
combine two equivalent points

drbenvincent committed:
update outdated text

drbenvincent committed:
Add pymc-marketing reference and notebook explanation

Added a citation for pymc-marketing to references.bib and updated the graded intervention time series notebook to explain the use of transformation functions from pymc-marketing for modeling temporal and intensity dynamics.

drbenvincent committed:
Revise HAC standard errors explanation in notebook

Clarified the rationale for HAC standard errors, improved the explanation of autocorrelation and heteroskedasticity, and streamlined the discussion of advantages and tradeoffs. Updated ARIMAX section reference for consistency.

drbenvincent committed:
Move import and setup code cell in notebook

juanitorduz commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+        # Fit OLS with HAC standard errors
+        if hac_maxlags is None:
+            n = len(y)
+            hac_maxlags = int(np.floor(4 * (n / 100) ** (2 / 9)))
```
where does this come from? any reference?

juanitorduz commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    base_formula: str,
+    saturation_type: Optional[str],
+    saturation_grid: Optional[Dict[str, List[float]]],
+    adstock_grid: Optional[Dict[str, List[float]]],
```
Again, remove `Optional` and use ` | ` instead

juanitorduz replied:
and all over the PR XD

juanitorduz commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    metric: str = "rmse",
+    error_model: str = "hac",
+    arima_order: Optional[Tuple[int, int, int]] = None,
+) -> Dict[str, Any]:
```
also use `dict` as we are now over Python 3.9 (we do not need to use these upper case type hints)

juanitorduz commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    adstock_bounds: Optional[Dict[str, Tuple[float, float]]],
+    initial_params: Optional[Dict[str, float]] = None,
+    coef_constraint: str = "nonnegative",
+    hac_maxlags: Optional[int] = None,
```
Rm Optionals

juanitorduz commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+            elif saturation_type == "michaelis_menten":
+                saturation = MichaelisMentenSaturation(**sat_kwargs)
+            else:
+                raise ValueError(f"Unknown saturation type: {saturation_type}")
```
I see this many times, can we simply write a helpter function with this logic?

drbenvincent replied:
Good call! Extracted a `create_saturation()` factory function in `transforms.py` backed by a `SATURATION_TYPES` registry dict. The 3 identical if/elif/else dispatch blocks in `transform_optimization.py` are now one-liners.

Done in 97c1f259.

juanitorduz commented on causalpy/transforms.py:
```diff
@@ -0,0 +1,494 @@
+    name: str
+    saturation: Optional[SaturationTransform] = None
+    adstock: Optional[AdstockTransform] = None
+    lag: Optional[LagTransform] = None
```
Dont use Optionals

drbenvincent replied:
Already done — the `Treatment` dataclass now uses `SaturationTransform | None`, `AdstockTransform | None`, `LagTransform | None` (no `Optional` usage remains).

juanitorduz commented on causalpy/transforms.py:
```diff
@@ -0,0 +1,494 @@
+        return {"lam": self.lam}
+
+
+class MichaelisMentenSaturation(SaturationTransform):
```
Why don't we directly take these from PyMC-Marketing?

juanitorduz commented on causalpy/pymc_models.py:
```diff
@@ -923,3 +923,1051 @@ class initialisation.
+        adstock_config: Optional[Dict] = None,
+        saturation_config: Optional[Dict] = None,
+        coef_constraint: str = "unconstrained",
+        sample_kwargs: Optional[Dict[str, Any]] = None,
```
Dont'e use Optionals

juanitorduz commented on causalpy/pymc_models.py:
```diff
@@ -923,3 +923,1051 @@ class initialisation.
+        self.treatment_names = None
+        self.n_treatments = None
+
+    def priors_from_data(self, X, y) -> Dict[str, Any]:
```
Missing type hints (also, use `dict`)

juanitorduz commented on causalpy/skl_models.py:
```diff
@@ -14,8 +14,10 @@
+from typing import Optional, Tuple
 
 import numpy as np
+import statsmodels.api as sm
```
Again, suggest using https://github.com/py-econometrics/pyfixest

juanitorduz reviewed (changes requested):
This is an amazing PR! I had no idea about this literature. I left some initial comments around code style. In a next round I could go deeper into the logic 💪 

Really cool stuff !

drbenvincent committed:
Merge main into tf-its and fix lint/type issues

Resolve merge conflicts between tf-its branch and main:
- Keep both main's and PR's additions in __init__.py, pyproject.toml,
  references.bib, notebooks/index.md
- Combine pymc_models.py: main's BayesianBasisExpansionTimeSeries and
  StateSpaceTimeSeries + PR's TransferFunction classes
- Accept main's deletion of interrogate_badge.svg
- Use main's pymc-marketing>=0.13.1 version constraint
- Modernize type annotations (Optional/Dict -> X | None/dict)
- Fix ruff lint issues (C408, B905, SIM105, UP)
- Fix mypy errors: add None guards, proper type annotations,
  use ConvMode enum, fix variable shadowing, np.asarray wrapping

Made-with: Cursor

drbenvincent committed:
Update copyright year to 2026 in test_skl_models_transfer_function.py

Made-with: Cursor

drbenvincent committed:
Add effect_summary stub to satisfy BaseExperiment ABC contract

GradedInterventionTimeSeries now implements the required effect_summary
abstract method with a NotImplementedError, directing users to the
.effect() method for counterfactual effect estimation.

Made-with: Cursor

drbenvincent commented:
## Branch sync and fixes to bring PR to green

This PR was 222 commits behind `main` and had merge conflicts. Here's what was done to bring it current:

### Merge conflict resolution (7 files)
- **`causalpy/__init__.py`**: Kept both main's additions (StaggeredDiD, extract_lift_for_mmm, variable_selection_priors) and PR's (GradedInterventionTimeSeries, transforms)
- **`causalpy/pymc_models.py`**: Combined main's new classes (BayesianBasisExpansionTimeSeries, StateSpaceTimeSeries) with PR's TransferFunction classes
- **`pyproject.toml`**: Adopted main's `pymc-marketing>=0.13.1` version constraint
- **`docs/source/references.bib`**: Merged bibliography entries from both branches
- **`docs/source/notebooks/index.md`**: Kept both Graded ITS and Comparative ITS notebook sections
- **`causalpy/tests/test_integration_pymc_examples.py`**: Removed stale `try/except` superseded by main's `contextlib.suppress`
- **`docs/source/_static/interrogate_badge.svg`**: Accepted main's deletion

### Lint and type fixes
- Modernized type annotations: `Optional[X]` → `X | None`, `Dict` → `dict` (ruff UP rules)
- Fixed ruff issues: `dict()` → dict literals (C408), `zip()` without `strict=` (B905), `try/except/pass` → `contextlib.suppress` (SIM105)
- Fixed ~50 mypy errors across PR files: added None guards for `idata` access, `np.asarray()` wrapping for pandas `.values` calls, proper base-class type declarations for transform variables, `ConvMode` enum usage
- Updated copyright header to 2026

### ABC compliance
- Added `effect_summary()` stub to `GradedInterventionTimeSeries` to satisfy the `BaseExperiment` abstract method contract introduced on `main` after this PR was created. Raises `NotImplementedError` directing users to `.effect()` for now.

All checks now pass: tests (3.11, 3.14), pre-commit.ci, docs build, packaging.

drbenvincent committed:
Address type hint review feedback from PR #548

- Add missing type hints for saturation, adstock, lag params in
  _fit_ols_with_transforms
- Use Literal["hac", "arimax"] for error_model params instead of str
- Replace Optional[Any] with Any | None in docstrings
- Replace Dict[str, Prior] with dict[str, Prior] in docstrings
- Add type annotations to treatment_names and n_treatments attributes
- Add parameter type hints to priors_from_data (X, y) in TF classes
- Add full type hints to build_model in both TF classes

Made-with: Cursor

drbenvincent commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    base_formula: str,
+    saturation,
+    adstock,
+    lag=None,
```
Done — added `SaturationTransform | None`, `AdstockTransform | None`, and `LagTransform | None` type hints to `saturation`, `adstock`, and `lag` parameters in `_fit_ols_with_transforms`.

drbenvincent commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    adstock,
+    lag=None,
+    hac_maxlags: Optional[int] = None,
+    error_model: str = "hac",
```
Good call! Changed `error_model: str` to `error_model: Literal["hac", "arimax"]` across all three functions in this module. (Note: used `"arimax"` rather than `"arima"` to match the actual valid values in the code.)

drbenvincent commented on causalpy/transform_optimization.py:
```diff
@@ -0,0 +1,602 @@
+    saturation,
+    adstock,
+    lag=None,
+    hac_maxlags: Optional[int] = None,
```
Already addressed — all `Optional[X]` usages were replaced with `X | None` in a previous round of changes.

drbenvincent commented on causalpy/pymc_models.py:
```diff
@@ -923,3 +923,1051 @@ class initialisation.
+        saturation_type: Optional[str] = None,
+        adstock_config: Optional[Dict] = None,
+        saturation_config: Optional[Dict] = None,
+        coef_constraint: str = "unconstrained",
```
Already done — no `Optional` remains in this class. All replaced with `X | None` syntax.

drbenvincent commented on causalpy/pymc_models.py:
```diff
@@ -923,3 +923,1051 @@ class initialisation.
+
+        return priors
+
+    def build_model(self, X, y, coords, treatment_data):
```
Done — `build_model` in `TransferFunctionARRegression` now has full type hints matching the linear regression variant: `X: xr.DataArray`, `y: xr.DataArray`, `coords: dict[str, Any]`, `treatment_data: xr.DataArray`, `-> None`.

drbenvincent committed:
Extract create_saturation() factory to eliminate repeated dispatch logic

Addresses review feedback from @juanitorduz on PR #548: the
saturation_type → SaturationTransform if/elif/else block was duplicated
3 times in transform_optimization.py. Replaced with a single
create_saturation() factory function in transforms.py backed by a
SATURATION_TYPES registry dict.

Made-with: Cursor

drbenvincent committed:
Merge origin/main into tf-its

Made-with: Cursor

drbenvincent committed:
Fix optional pymc-marketing handling for CI

Made-with: Cursor

drbenvincent committed:
Handle geometric adstock API changes

Made-with: Cursor

drbenvincent committed:
Support xtensor geometric adstock calls

Made-with: Cursor

drbenvincent committed:
Support xtensor Fourier seasonality calls

Handle newer pymc-marketing Fourier components by retrying seasonality inputs with obs_ind xtensor dims, so the PR's BSTS and transfer-function PyMC tests keep working across dependency versions.

Made-with: Cursor

drbenvincent committed:
Handle xtensor saturation and seasonality APIs

Detect newer pymc-marketing transformer variants before calling them so BSTS seasonality and transfer-function saturation paths work without duplicate variable registration or missing dims errors.

Made-with: Cursor

drbenvincent committed:
Handle xtensor trend component calls

Apply the same upfront xtensor detection to BSTS time components so newer pymc-marketing trend APIs receive obs_ind dims without regressing older transformer signatures.

Made-with: Cursor

drbenvincent committed:
Cover xtensor detection fallback

Exercise the helper path that falls back to bytecode names when source inspection is unavailable so patch coverage clears the Codecov threshold.

Made-with: Cursor

drbenvincent committed:
Handle xtensor BSTS pymc-marketing components

Wrap shared BayesianBasisExpansionTimeSeries trend and seasonality components so newer pymc-marketing xtensor APIs work when the package is installed, without tying the fix to TF-ITS-specific transformer code.

Made-with: Cursor

drbenvincent committed:
Merge shared pymc-marketing compatibility branch

Made-with: Cursor

drbenvincent committed:
Merge origin/main after compatibility PR

Made-with: Cursor

drbenvincent committed:
Merge branch 'main' into tf-its

drbenvincent commented:
## Automated triage

**Recommendation:** `review:high` — no decision gate identified.

**Why:**
- Major new feature: Transfer-Function ITS with graded interventions, saturation/adstock transforms, new experiment class (1887 lines).
- New dependency: `pymc-marketing>=0.7.0` added to `pyproject.toml`.
- Large new modules: `graded_intervention_its.py`, `transform_optimization.py`, new model code in `pymc_models.py` (1132 lines) and `skl_models.py`.
- Extensive tests: 1873 lines in `test_transfer_function_its.py`, 633 in `test_skl_models_transfer_function.py`.
- `codecov/patch` FAILED. PR is conflicted (DIRTY). Review already has requested changes.

**Review focus:**
1. Review the new `pymc-marketing` dependency: is `>=0.7.0` the correct minimum version?
2. Verify the transform system (Saturation, Adstock, Lag, Treatment dataclasses) is well-designed for future extensibility.
3. Check the diagnostic plots (ACF/PACF, IRF) and Ljung-Box test integration.

**Confidence:** high
